sound/soc/sdw_utils/soc_sdw_utils.c

```diff
@@ -0,0 +1,990 @@
+				.num_widgets = ARRAY_SIZE(generic_jack_widgets),
+			},
+		},
+		.dai_num = 1,
+	},
+	{
+		.part_id = 0x4243,
+		.codec_name = "cs42l43-codec",
+		.count_sidecar = asoc_sdw_bridge_cs35l56_count_sidecar,
+		.add_sidecar = asoc_sdw_bridge_cs35l56_add_sidecar,
+		.dais = {
+			{
+				.direction = {true, false},
+				.dai_name = "cs42l43-dp5",
+				.dai_type = SOC_SDW_DAI_TYPE_JACK,
+				.dailink = {SOC_SDW_JACK_OUT_DAI_ID, SOC_SDW_UNUSED_DAI_ID},
+				.rtd_init = asoc_sdw_cs42l43_hs_rtd_init,
+				.controls = generic_jack_controls,
+				.num_controls = ARRAY_SIZE(generic_jack_controls),
+				.widgets = generic_jack_widgets,
+				.num_widgets = ARRAY_SIZE(generic_jack_widgets),
+			},
+			{
+				.direction = {false, true},
+				.dai_name = "cs42l43-dp1",
+				.dai_type = SOC_SDW_DAI_TYPE_MIC,
+				.dailink = {SOC_SDW_UNUSED_DAI_ID, SOC_SDW_DMIC_DAI_ID},
+				.rtd_init = asoc_sdw_cs42l43_dmic_rtd_init,
+				.widgets = generic_dmic_widgets,
+				.num_widgets = ARRAY_SIZE(generic_dmic_widgets),
+			},
+			{
+				.direction = {false, true},
+				.dai_name = "cs42l43-dp2",
+				.dai_type = SOC_SDW_DAI_TYPE_JACK,
+				.dailink = {SOC_SDW_UNUSED_DAI_ID, SOC_SDW_JACK_IN_DAI_ID},
+			},
+			{
+				.direction = {true, false},
+				.dai_name = "cs42l43-dp6",
+				.dai_type = SOC_SDW_DAI_TYPE_AMP,
+				.dailink = {SOC_SDW_AMP_OUT_DAI_ID, SOC_SDW_UNUSED_DAI_ID},
+				.init = asoc_sdw_cs42l43_spk_init,
+				.rtd_init = asoc_sdw_cs42l43_spk_rtd_init,
+				.controls = generic_spk_controls,
+				.num_controls = ARRAY_SIZE(generic_spk_controls),
+				.widgets = generic_spk_widgets,
+				.num_widgets = ARRAY_SIZE(generic_spk_widgets),
+				.quirk = SOC_SDW_CODEC_SPKR | SOC_SDW_SIDECAR_AMPS,
+			},
+		},
+		.dai_num = 4,
+	},
+	{
+		.part_id = 0xaaaa, /* generic codec mockup */
+		.version_id = 0,
+		.dais = {
+			{
+				.direction = {true, true},
+				.dai_name = "sdw-mockup-aif1",
+				.dai_type = SOC_SDW_DAI_TYPE_JACK,
+				.dailink = {SOC_SDW_JACK_OUT_DAI_ID, SOC_SDW_JACK_IN_DAI_ID},
+			},
+		},
+		.dai_num = 1,
+	},
+	{
+		.part_id = 0xaa55, /* headset codec mockup */
+		.version_id = 0,
+		.dais = {
+			{
+				.direction = {true, true},
+				.dai_name = "sdw-mockup-aif1",
+				.dai_type = SOC_SDW_DAI_TYPE_JACK,
+				.dailink = {SOC_SDW_JACK_OUT_DAI_ID, SOC_SDW_JACK_IN_DAI_ID},
+			},
+		},
+		.dai_num = 1,
+	},
+	{
+		.part_id = 0x55aa, /* amplifier mockup */
+		.version_id = 0,
+		.dais = {
+			{
+				.direction = {true, true},
+				.dai_name = "sdw-mockup-aif1",
+				.dai_type = SOC_SDW_DAI_TYPE_AMP,
+				.dailink = {SOC_SDW_AMP_OUT_DAI_ID, SOC_SDW_AMP_IN_DAI_ID},
+			},
+		},
+		.dai_num = 1,
+	},
+	{
+		.part_id = 0x5555,
+		.version_id = 0,
+		.dais = {
+			{
+				.dai_name = "sdw-mockup-aif1",
+				.direction = {false, true},
+				.dai_type = SOC_SDW_DAI_TYPE_MIC,
+				.dailink = {SOC_SDW_UNUSED_DAI_ID, SOC_SDW_DMIC_DAI_ID},
+			},
+		},
+		.dai_num = 1,
+	},
+};
+EXPORT_SYMBOL_NS(codec_info_list, SND_SOC_SDW_UTILS);
+
```

Comment:
Feels a bit weird having all this stuff at the ASoC level, is this not a specific machine driver implementation, one that is shared between AMD and Intel now, but its not really generic, it is data for a specific implementation. That said I am not sure exactly where it should live, maybe the sof level? Are we calling this the sof machine driver implementation?

Reply:
@charleskeepax: Soon we are going to add Legacy (NO DSP) machine driver which will refer same codec_info_list structure and helpers pulled in sdw_utils folder will be utilized by our legacy machine driver.  The reason beyond pulling the common helpers/structures at one place , so that Intel & AMD generic machine driver code base going to refer these common structures and helper functions.

Reply:
what makes you think this is implementation-specific @charleskeepax?

This table augments the basic dai descriptors by adding SDCA-type qualifiers that help create dailinks based on an ACPI description. That ACPI description follows the DisCo spec, it's not machine or SOC-specific.

I think the concept maps well for all SDCA topologies, we have dedicated streaming IT/OT for different functionality, I don't see currently a risk of a DAI/terminal being used for machine-specific routing. If you want to access to jack output or an ultrasound render terminal you are required to rely on an SDCA-defined terminal.

Reply:
Now to be clear I don't have a mega-strong objection here, if Mark is happy to merge it I don't really mind it going in as is and it is not like putting this in ASoC prevents someone doing something else either.

But the code is very clearly one particular way of implementing things. Even the "acpi" bits are not some standardised acpi thing, they are just a bunch of tables contained in the machine driver. Basically the information encoding what the DAI links look like is partitioned between the acpi structs and the codec_info structs, but there is no reason to believe every machine driver would always want to make those same choices of how to partition that information. Even more so when you get to things like the init and exit callbacks, which even more explicitly depend on a specific implementation.

The way it is laid out does likely map reasonably well to SDCA, as they acpi structs fill in mostly what is missing from the disco description, in that case most of the information in the codec_info side should not be required as at that point it is either coming from the definition of the topology for the function or from the disco description of it.

Reply:
agree, this list will probably disappear at some point and will be replaced by properties read from DisCo. There's no intention to make the integration of every codec in this list a requirement, it's more of a stop-gap solution until we have a better one.